_drafts/2023-04-15-functional-error-handling-in-kotlin.md

```diff
@@ -0,0 +1,811 @@
+    val job: Job = throw NoSuchElementException("Job not found")
+    return try {
+        job.salary.value
+    } catch (e: Exception) {
+        0.0
+    }
+}
+```
+
+As expected, we obtained a program that crashes with the same exception. However, if we move the retrieval of the job inside the `try-catch` block, we get the following code:
+
+```kotlin
+fun retrieveSalary(id: JobId): Double {
+    return try {
+        val job = jobs.findById(id)
+        job.salary.value
+    } catch (e: Exception) {
+        0.0
+    }
+}
+```
+
+If we execute this code in our program, we obtain a different result than the previous execution. In fact, the exception is now caught by the `try-catch` block, and the program generates the following output.
+
+```text
+The salary of the job 42 is 0.0
+```
+
+We've just proven that exceptions don't follow the substitution principle. In other words, exceptions are not referentially transparent and, for this reason, are considered a side effect. In other words, **an expression throwing exceptions can't be reasoned about without a context of execution, aka the code around the expression**. So, when we use the expression, we also lose the locality of reasoning and add a lot of cognitive loads to understand the code.
+
+Moreover, there is one more aspect we would like to change about exceptions. Let's take the signature of the `retrieveSalary` method:
+
+```kotlin
+fun retrieveSalary(id: JobId): Double
+```
+
+We expect the method to take a job id as input and return a salary as a `Double`. No reference to the exception is present in the signature. **As developers, we want the compilers to help us avoid errors**. However, in this case, we're not aware that the method can throw an exception, and the compiler can not help us in any way. The only place we become aware of the exception is during runtime execution, which is a bit late.
+
+Somebody can say that the JVM also has checked exceptions and that we can use them to avoid the problem. However, **Kotlin doesn't have checked exceptions**. Let's try to act as it has them. If a method declares to throw a checked exception, the compiler will force us to handle it. But, **checked exceptions don't work well with higher-order functions**, which are fundamental to functional programming. In fact, if we want to use a higher-order function together with checked exceptions, we need to declare the exception in the signature of the lambda function, which is not feasible. Take the `map` function of any collection type:
+
+```kotlin
+fun <A, B> map(list: List<A>, f: (A) -> B): List<B>
+```
+
+As we might guess, using checked exceptions for the function `f` is impossible since it must stay generic. The only possible way is to add some generic exception to the function's signature, such as an `Exception`, which is useless.
+
+So, we understood that we need a better approach to handle errors, at least in functional programming. Let's see how we can do it.
```

Comment:
I like the explanations of these last few paragraphs - they summarize why we don't (usually) have checked exceptions in FP very well. ZIO is the only counterexample and even that is just partial.